### PR TITLE
Add clients

### PR DESCRIPTION
**Pull request by gregcusack**
#### Summary of Changes
1) Create client accounts
2) Create and deploy client secrets, services, replica sets
3) Fix nits from jon: https://github.com/anza-xyz/validator-lab/pull/40

### CODE REVIEW

File: src/genesis.rs [hunk 1]
```diff
@@ -179,6 +210,83 @@ impl Genesis {
...
+        executable_path: &PathBuf,
+    ) -> Result<Child, Box<dyn Error>> {
+        let child = Command::new(executable_path)
+            .args(args)
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .spawn()?;
+
+        Ok(child)
+    }
```
Comment (joncinque) [inline src/genesis.rs:288]: nit: this function name is pretty confusing, since it's very generically just running a command. I'd say to either keep the logic and rename it to `spawn_command`, or add more logic to build up the args and all that

File: src/main.rs [hunk 2]
```diff
@@ -237,10 +236,67 @@ fn parse_matches() -> clap::ArgMatches {
...
-                }),
+        )
+        // Client Config
+        .arg(
+            Arg::with_name("number_of_clients")
+                .long("num-clients")
+                .short('c')
+                .takes_value(true)
+                .default_value("0")
+                .help("Number of clients ")
```
Comment (joncinque) [inline src/main.rs:247]: micro-nit
```suggestion
                .help("Number of clients")
```

File: src/main.rs [hunk 3]
```diff
@@ -237,10 +236,67 @@ fn parse_matches() -> clap::ArgMatches {
...
+                .takes_value(true)
+                .help("Client Config. Optional: Specify an exact node to send transactions to. use: --target-node <Pubkey>.
+                Not supported yet. TODO..."),
+        )
+        .arg(
+            Arg::with_name("duration")
+                .long("duration")
+                .takes_value(true)
+                .default_value("7500")
+                .help("Client Config. Seconds to run benchmark, then exit; default is forever use: --duration <SECS>"),
```
Comment (joncinque) [inline src/main.rs:292]: Given, you've already added a default, you can also add `.value_name("SECS")` and simplify this message to
```suggestion
                .help("Client Config. Seconds to run benchmark, then exit"),
```

File: src/main.rs [hunk 4]
```diff
@@ -237,10 +236,67 @@ fn parse_matches() -> clap::ArgMatches {
...
+        )
+        .arg(
+            Arg::with_name("num_nodes")
+                .long("num-nodes")
+                .short('N')
```
Comment (joncinque) [inline src/main.rs:297]: This name is unfortunately a little unclear, how about `--client-wait-for-n-nodes` or `--wait-for-nodes`?

File: src/main.rs [hunk 5]
```diff
@@ -237,10 +236,67 @@ fn parse_matches() -> clap::ArgMatches {
...
+                .takes_value(true)
+                .help("Client Config. Optional: Wait for NUM nodes to converge: --num-nodes <NUM> "),
```
Comment (joncinque) [inline src/main.rs:299]: with `.value_name("NUM")` you can also simplify this help text
```suggestion
                .help("Client Config. Optional: Wait for NUM nodes to converge"),
```

File: src/main.rs [hunk 6]
```diff
@@ -237,10 +236,67 @@ fn parse_matches() -> clap::ArgMatches {
...
+                User can optionally provide extraArgs that are transparently
+                supplied to the client program as command line parameters.
+                For example,
+                    --bench-tps-args 'tx-count=5000 thread-batch-sleep-ms=250'
+                This will start bench-tps clients, and supply '--tx-count 5000 --thread-batch-sleep-ms 250'
+                to the bench-tps client."),
+        )
+        .arg(
+            Arg::with_name("target_node")
+                .long("target-node")
```
Comment (joncinque) [inline src/main.rs:282]: nit: how about `--client-target-node` instead to make it clearer?

File: src/main.rs [hunk 7]
```diff
@@ -237,10 +236,67 @@ fn parse_matches() -> clap::ArgMatches {
...
+                .takes_value(true)
+                .help("Client Config. Optional: Specify an exact node to send transactions to. use: --target-node <Pubkey>.
+                Not supported yet. TODO..."),
+        )
+        .arg(
+            Arg::with_name("duration")
+                .long("duration")
```
Comment (joncinque) [inline src/main.rs:289]: nit: How about calling this `--client-duration-seconds` to be safe?

File: src/main.rs [hunk 8]
```diff
@@ -310,6 +366,29 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     let num_validators = value_t_or_exit!(matches, "number_of_validators", usize);
     let num_rpc_nodes = value_t_or_exit!(matches, "number_of_rpc_nodes", usize);
+    let client_config = ClientConfig {
+        num_clients: value_t_or_exit!(matches, "number_of_clients", usize),
+        client_type: matches
+            .value_of("client_type")
+            .unwrap_or_default()
```
Comment (joncinque) [inline src/main.rs:373]: I think with `default_value` you can just do a plain `unwrap`, since `unwrap_or_default` will give an empty string

File: src/main.rs [hunk 9]
```diff
@@ -310,6 +366,29 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+            .to_string(),
+        client_to_run: matches
+            .value_of("client_to_run")
+            .unwrap_or_default()
```
Comment (joncinque) [inline src/main.rs:377]: same here, I think `unwrap()` will be better

File: src/main.rs [hunk 10]
```diff
@@ -310,6 +366,29 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+            .to_string(),
+        bench_tps_args: parse_and_format_bench_tps_args(matches.value_of("bench_tps_args")),
+        target_node: match matches.value_of("target_node") {
+            Some(s) => match s.parse::<Pubkey>() {
+                Ok(pubkey) => Some(pubkey),
+                Err(e) => return Err(format!("failed to parse pubkey in target_node: {e}").into()),
+            },
+            None => None,
+        },
```
Comment (joncinque) [inline src/main.rs:386]: Feel free to use `clap_utils::input_parsers::pubkey_of` to avoid this annoyance

File: src/main.rs [hunk 11]
```diff
@@ -558,7 +658,7 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     kub_controller
         .deploy_secret(bootstrap_validator.secret())
         .await?;
-    info!("Deployed Bootstrap Secret");
+    info!("Bootstrap Secret deployed");
```
Comment (joncinque) [inline src/main.rs:661]: micro-nit: I actually prefer the new active style more that you're using in the other logs, but it's up to you
```suggestion
    info!("Deployed Bootstrap Secret");
```

File: src/main.rs [hunk 12]
```diff
@@ -688,90 +785,122 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
 
         // wait for at least one rpc node to deploy
-        loop {
-            let mut ready = false;
+        'outer: loop {
             for rpc_name in &rpc_nodes {
                 if kub_controller.is_replica_set_ready(rpc_name).await? {
-                    ready = true;
-                    break;
+                    break 'outer;
```
Comment (joncinque) [inline src/main.rs:804]: :muscle:

File: src/main.rs [hunk 13]
```diff
@@ -688,90 +785,122 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-            info!("no rpc nodes ready yet");
+            info!("RPC Nodes not ready yet");
             tokio::time::sleep(std::time::Duration::from_secs(1)).await;
         }
     }
 
-    if num_validators == 0 {
-        info!("No validators to deploy. Returning");
-        return Ok(());
+    if num_validators > 0 {
```
Comment (joncinque) [inline src/main.rs:813]: Highly recommend "hide whitespace" to review this part :sweat_smile:

**Review by joncinque — COMMENTED** [review]
Looking really good! Just small things

File: src/genesis.rs [hunk 14]
```diff
@@ -92,6 +95,34 @@ impl std::fmt::Display for GenesisFlags {
     }
 }
 
+fn append_client_accounts_to_file(
+    in_file: &PathBuf,  //bench-tps-i.yml
+    out_file: &PathBuf, //client-accounts.yml
```
[the remark this answers is not in the document]
Comment (gregcusack) [reply src/genesis.rs:100]: great point lol

File: src/genesis.rs [hunk 15]
```diff
@@ -179,6 +210,83 @@ impl Genesis {
...
+
+        for child in children? {
+            let output = child.wait_with_output()?;
+            if !output.status.success() {
+                return Err(output.status.to_string().into());
+            }
+        }
+
+        let progress_bar = new_spinner_progress_bar();
+        progress_bar.set_message(format!("{WRITING}Writing client accounts..."));
```
[the remark this answers is not in the document]
Comment (gregcusack) [reply src/genesis.rs:266]: it is for combining all `bench-tps-{i}.yml` accounts into `client-accounts.yml` for genesis. it can take a while if there are a lot of client accounts. could def put in for whole function.

File: src/main.rs (hunk 11 above)
Comment (gregcusack) [reply src/main.rs:661]: ahh ya i went back and forth on this. will update

File: src/docker.rs [hunk 16]
```diff
@@ -208,6 +215,40 @@ WORKDIR /home/solana
...
+        validator_type: &ValidatorType,
+    ) -> Result<String, Box<dyn Error>> {
+        match validator_type {
+            ValidatorType::Client(index) => {
+                let bench_tps_path =
+                    solana_root_path.join(format!("config-k8s/bench-tps-{index}.yml"));
+                if bench_tps_path.exists() {
+                    Ok(format!(
+                        r#"
+COPY --chown=solana:solana ./config-k8s/bench-tps-{index}.yml /home/solana/client-accounts.yml
```
[the remark this answers is not in the document]
Comment (gregcusack) [reply src/docker.rs:230]: ahh we can't really do this since `bench_tps_path` is the full path on the host. whereas `./config-k8s/bench-tps-{index}.yml` is the path in the container.

File: src/genesis.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (joncinque) [reply src/genesis.rs]: Ah gotcha, thanks!

**Review by yihau — APPROVED** [review]
sorry for the very late review 😢 looks good to me! let's do it!

File: src/docker.rs [hunk 17]
```diff
@@ -196,8 +197,10 @@ COPY --chown=solana:solana ./{solana_build_directory}/version.yml /home/solana/
 ENV PATH="/home/solana/.cargo/bin:${{PATH}}"
```
Comment (joncinque) [inline src/docker.rs:197]: Did you mean to revert the changes to not use `.cargo`?

File: README.md [hunk 18]
```diff
@@ -63,6 +63,12 @@ cargo run --bin cluster --
     # kubernetes config
     --cpu-requests <cores>
     --memory-requests <memory>
+    # deploy with clients
+    -c <num-clients>
+    --client-type <client-type e.g. tpu-client>
+    --client-to-run <type-of-client e.g. bench-tps>
+    --num-nodes <wait-for-N-nodes-to-converge-before-starting-client>
```
Comment (joncinque) [inline README.md:70]:
```suggestion
    --client-wait-for-n-nodes <wait-for-N-nodes-to-converge-before-starting-client>
```

File: src/startup_scripts.rs [hunk 19]
```diff
@@ -1037,7 +1030,6 @@ bench-tps)
   clientCommand="\
     solana-bench-tps \
       --sustained \
-      --threads $threadCount \
```
Comment (joncinque) [inline src/startup_scripts.rs]: Did you mean to remove this?

**Review by joncinque — COMMENTED** [review]
Just a couple of last things

File: src/startup_scripts.rs (hunk 19 above)
Comment (gregcusack) [reply src/startup_scripts.rs]: ya i did. because the way it was setup, you couldn't actually set the threads with cli flags. it would just take the number of cores and if number of cores was greater than 4, it would set to 4. And the default thread count in bench-tps is 4 anyway. So I took it out. now you can pass in `threads=4` into `--bench-tps-args`

File: src/docker.rs (hunk 17 above)
Comment (gregcusack) [reply src/docker.rs:197]: what do you mean? I am using `.cargo`. All the built binaries get stored in `/home/solana/.cargo/bin` and then I set the `PATH` here, so you can just run: `solana-bench-tps`, `agave-validator`, etc

Comment (joncinque) [reply src/docker.rs:197]: Oh ok, earlier in the PR you had changed it to go to `/home/solana/bin` instead, and then the last commit went back to using `/home/solana/.cargo/bin`. So is the final decision to use the `.cargo` dir instead?

File: src/startup_scripts.rs (hunk 19 above)
Comment (joncinque) [reply src/startup_scripts.rs]: gotcha, thanks!

File: src/docker.rs (hunk 17 above)
Comment (gregcusack) [reply src/docker.rs:197]: oh ya my bad. not sure exactly what I was thinking and why I switched it back. I must have misread a comment. I am going to switch it back to `/home/solana/bin`

**Review by joncinque — APPROVED** [review]
Thanks for addressing everything, looks great!
